### PR TITLE
ORCOMP-304 Implemented custom DateTime formatter/parser to solve copy/paste issues

### PR DESCRIPTION
**Pull request by GeertvanHorrik**

### CODE REVIEW

File: src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs [hunk 1]
```diff
@@ -278,9 +278,10 @@ private void OnCopyButtonClick(object sender, RoutedEventArgs e)
         {
             DatePickerIcon.SetCurrentValue(ToggleButton.IsCheckedProperty, false);
 
-            if (Value != null)
+            var value = Value;
+            if (value != null)
             {
-                Clipboard.SetText(Value.Value.ToString(Format), TextDataFormat.Text);
+                Clipboard.SetText(DateTimeFormatter.Format(value.Value, _formatInfo), TextDataFormat.Text);
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs]: Should we check for value.Value == null?

File: src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs [hunk 2]
```diff
@@ -372,9 +374,9 @@ private void ApplyFormat()
 
             EnableOrDisableYearConverterDependingOnFormat();
 
-            NumericTBDay.SetCurrentValue(NumericTextBox.FormatProperty, GetFormat(_formatInfo.DayFormat.Length));
-            NumericTBMonth.SetCurrentValue(NumericTextBox.FormatProperty, GetFormat(_formatInfo.MonthFormat.Length));
-            NumericTBYear.SetCurrentValue(NumericTextBox.FormatProperty, GetFormat(_formatInfo.YearFormat.Length));
+            NumericTBDay.SetCurrentValue(NumericTextBox.FormatProperty, NumerFormatHelper.GetFormat(_formatInfo.DayFormat.Length));
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs]: Typo, should be *NumberFormatHelper*

File: src/Orc.Controls/Orc.Controls.Shared/Helpers/DateTimeParser.cs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+                }
+            }
+
+            try
+            {
+                return new DateTime(year, month, day, hour, minute, second);
+            }
+            catch (Exception exc)
+            {
+                return ThrowOnError<FormatException>("Invalid value. It's not possible to build new DateTime object", exc);
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.Shared/Helpers/DateTimeParser.cs]: why not `throw Log.CreateException<FormatException>("Invalid value. It's not possible to build new DateTime object")`?

File: src/Orc.Controls/Orc.Controls.Shared/Helpers/NumerFormatHelper.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Orc.Controls
+{
+    using Catel;
+    using Catel.Logging;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public static class NumerFormatHelper
```
Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.Shared/Helpers/NumerFormatHelper.cs]: Typo, should be NumberFormatHelper

File: src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs (hunk 1 above)
Comment (michu) [inline src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs]: We're checking already in line 281.

File: src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs (hunk 2 above)
Comment (michu) [inline src/Orc.Controls/Orc.Controls.Shared/Controls/DatePicker/Views/DatePickerControl.xaml.cs]: Yeah, typo, in my language 'Numer' is valid form :) Thx. Fixed.

File: src/Orc.Controls/Orc.Controls.Shared/Helpers/NumerFormatHelper.cs (hunk 4 above)
Comment (michu) [inline src/Orc.Controls/Orc.Controls.Shared/Helpers/NumerFormatHelper.cs]: Fixed.

File: src/Orc.Controls/Orc.Controls.Shared/Helpers/DateTimeParser.cs (hunk 3 above)
Comment (michu) [inline src/Orc.Controls/Orc.Controls.Shared/Helpers/DateTimeParser.cs]: I'm using this ThrowOnError<TException> method to support TryParse and Parse methods like in standard .NET parsers. Throwing exceptions is expensive, that's why I throw or return false depending on throwOnError parameter. Should I add logging to ThrowOnError<TException> method ?

Comment (GeertvanHorrik) [inline src/Orc.Controls/Orc.Controls.Shared/Helpers/DateTimeParser.cs]: *if* throw on error is used, then indeed log it

**Pull request merged**